KAFKA-6738: Implement error handling for source and sink tasks

wicknicks opened the pull request:
This PR implements the features described in this KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-298%3A+Error+Handling+in+Connect

This PR changes the Connect framework to allow it to automatically deal with errors encountered while processing records in a Connector. The following behavior changes are introduced here:

**Retry on Failure**: Retry the failed operation a configurable number of times, with backoff between each retry.
**Task Tolerance Limits**: Tolerate a configurable number of failures in a task.

We also add the following ways to report errors, along with sufficient context to simplify the debugging process:

**Log Error Context**: The error information along with processing context is logged along with standard application logs.
**Dead Letter Queue**: Produce the original message into a Kafka topic (applicable only to sink connectors).

New **metrics** which will monitor the number of failures, and the behavior of the response handler are added.

The changes proposed here **are backward compatible**. The current behavior in Connect is to kill the task on the first error in any stage. This will remain the default behavior if the connector does not override any of the new configurations which are provided as part of this feature.

Testing: added multiple unit tests to test the retry and tolerance logic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

wicknicks committed:
KAFKA-6738: Implement error handling for source and sink tasks

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Add unit tests for RetryWithToleranceExecutor

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Changes for Java 8

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Add retry logic to task.poll() in WorkerSourceTask

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@mageshn, @rhauch can I get some early reviews for the error handling PR. Since the deadline for feature freeze is coming up soon, it would be better to get some early thoughts on how these changes are coming along. I'll adding more unit tests later today, with javadocs and style fixes coming up after that.

mjsax commented:
@wicknicks JIRA number should be KAFKA-6738 ?

wicknicks commented:
ah shoot, yes. thanks, @mjsax!

wicknicks committed:
KAFKA-6738: Add tests for DLQReporter

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Add test for error reporters

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Worker test with transformers and converters

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Make transform failure behavior more consistent

Signed-off-by: Arjun Satish <arjun@confluent.io>

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -32,10 +37,26 @@ public TransformationChain(List<Transformation<R>> transformations) {
     }
 
     public R apply(R record) {
+        return apply(record, NoopExecutor.INSTANCE, new ProcessingContext());
```
wondering if there was a reason to leave this default method?

wicknicks replied:
leads to fewer changes in the codebase, easier to merge larger PRs.

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/Operation.java:
```diff
@@ -0,0 +1,21 @@
+
+public interface Operation<V> {
+    V apply() throws Exception;
+}
```
nit : missing new line

wicknicks replied:
fixed this, but for some reason it's not showing up here.

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:
```diff
@@ -241,17 +290,48 @@ private boolean sendRecords() {
-            byte[] value = valueConverter.fromConnectData(record.topic(), record.valueSchema(), record.value());
+            RecordHeaders headers;
+            processingContext.setStage(Stage.HEADER_CONVERTER, headerConverter.getClass());
+            Result<RecordHeaders> headersResult = execute(() -> convertHeaderFor(record));
```
Looks like setting the stage, processing the result can all be handled in the execute method

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/NoopExecutor.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class NoopExecutor extends RetryWithToleranceExecutor {
```
Is this required just so that we don't change the unit tests? If so, it might be much better to have a special instance in the RetryWithToleranceExecutor itself?

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,120 @@
+
+    public ProcessingContext result(Result<?> result) {
+        this.result = result;
+        return this;
```
Look like a ProcessingContext builder method while it is not. Wouldn't it be better to keep this void

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,120 @@
+        return this;
+    }
+
+    public void setStage(Stage stage, Class<?> klass) {
```
doesn't look a great name for its behavior. perhaps something like currentContext

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceExecutor.java:
```diff
@@ -0,0 +1,215 @@
+    @Override
+    public <V> Result<V> execute(Operation<V> operation, ProcessingContext context) {
+        try {
+            switch (context.stage()) {
```
could we condense this mapping in an enum or a static map such that we can avoid all this logic here?

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -461,13 +475,43 @@ private WorkerTask buildWorkerTask(ConnectorConfig connConfig,
+
+        // check if topic for dead letter queue exists
+        String topic = connConfig.getString(DLQReporter.PREFIX + "." + DLQReporter.DLQ_TOPIC_NAME);
+        if (topic != null && topic.length() > 0) {
```
can use isEmpty() instead of checking length

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -461,13 +475,43 @@ private WorkerTask buildWorkerTask(ConnectorConfig connConfig,
+        // check if topic for dead letter queue exists
+        String topic = connConfig.getString(DLQReporter.PREFIX + "." + DLQReporter.DLQ_TOPIC_NAME);
+        if (topic != null && topic.length() > 0) {
+            DescribeTopicsResult topics = AdminClient.create(config.originals()).describeTopics(Collections.singleton(topic));
```
The general pattern for all these topics is createOrVerify.

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -472,14 +501,48 @@ public String toString() {
+            if (valueResult.success()) {
+                valueAndSchema = valueResult.result();
+            } else {
+                processingContext.report();
```
Isn't the report invoked on the executor?

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -472,14 +501,48 @@ public String toString() {
         return newConsumer;
     }
 
+    protected <V> Result<V> execute(Operation<V> operation) {
```
Can this method be moved to super class? Also would it be better to pass the success and failure operations as call back as well to the executor?

wicknicks replied:
in most cases, we simply need to skip records. not sure if that logic can be passed into the executor. it's the sink/source task that needs to execute that piece.

mageshn replied:
The executor could possibly completely handle success vs error and actions for error. This will help reduce lot of error handling code in the core classes

mageshn reviewed (commented):
Great changes @wicknicks . Have done a high level first pass.

wicknicks committed:
KAFKA-6738: Add metrics for monitoring error handling

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Address reviewer comments

Signed-off-by: Arjun Satish <arjun@confluent.io>

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -457,17 +474,66 @@ private WorkerTask buildWorkerTask(ConnectorConfig connConfig,
             return new WorkerSourceTask(id, (SourceTask) task, statusListener, initialState, keyConverter, valueConverter,
-                    headerConverter, transformationChain, producer, offsetReader, offsetWriter, config, metrics, loader, time);
+                    headerConverter, transformationChain, producer, offsetReader, offsetWriter, config, metrics, loader,
+                    time, sourceProcessingContext(id, connConfig, errorHandlingMetrics), operationExecutor);
```
Nit: the call to `sourceProcessingContext` gets buried. How about pulling that out into a local variable just before line 476 similarly to how this is done with `producer`.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -457,17 +474,66 @@ private WorkerTask buildWorkerTask(ConnectorConfig connConfig,
             return new WorkerSinkTask(id, (SinkTask) task, statusListener, initialState, config, metrics, keyConverter,
-                    valueConverter, headerConverter, transformationChain, loader, time);
+                    valueConverter, headerConverter, transformationChain, loader, time, sinkProcessingContext(id, connConfig, errorHandlingMetrics),
+                    operationExecutor);
```
Nit: pull out call to `sinkProcessingContext` and use a local variable (see above).

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -457,17 +474,66 @@ private WorkerTask buildWorkerTask(ConnectorConfig connConfig,
+                dlqReporter.configure(connConfig.originalsWithPrefix(DLQReporter.PREFIX));
+                reporters.add(dlqReporter);
+            } catch (Exception e) {
+                throw new ConnectException(e);
```
This wraps the `ConfigException` thrown on line 510. Is that intentional?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -472,14 +501,45 @@ public String toString() {
-            SchemaAndValue valueAndSchema = toConnectData(valueConverter, "value", msg, msg.value());
-            Headers headers = convertHeadersFor(msg);
+
+            processingContext.sinkRecord(msg);
```
The `sinkRecord` seems an odd name, since it's not a `SinkRecord`. How about `consumerRecord`?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -472,14 +501,45 @@ public String toString() {
+                keyAndSchema = keyResult.result();
+            } else {
+                continue;
+            }
```
This organization seems strange to declare a variable we might not need. Why not something like:
```
            processingContext.setCurrentContext(Stage.KEY_CONVERTER, keyConverter.getClass());
            Result<SchemaAndValue> keyResult = execute(() -> keyConverter.toConnectData(msg.topic(), msg.key()));
            if (keyResult.failed()) {
                continue;
            }
            SchemaAndValue keyAndSchema = keyResult.result();
``` 

I admit it's a minor thing, but it seems better to handle the exceptional case in the block and to leave the typical (error-free) case at the same level.

If you agree, then it should be changed everywhere this pattern is used.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -32,10 +37,26 @@ public TransformationChain(List<Transformation<R>> transformations) {
+        return apply(record, RetryWithToleranceExecutor.NOOP_EXECUTOR, new ProcessingContext());
+    }
+
+    public R apply(R record, OperationExecutor operationExecutor, ProcessingContext processingContext) {
```
How about adding a `initialize(OperationExecutor, ProcessingContext)` method to `TransformationChain`, so that the apply signature can remain the same? The WorkerSourceTask and WorkerSinkTask both pass in the same objects on every `apply` call, so this would simplify things a bit. It'd avoid changing the signature, and would make it a bit cleaner by having the fields default to the no-op executor and empty context.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -32,10 +37,26 @@ public TransformationChain(List<Transformation<R>> transformations) {
+            } else {
+                processingContext.report();
+                return null;
+            }
```
This pattern is a bit strange (see other comment). How about:
```
            // execute the operation
            Result<R> result = operationExecutor.execute(() -> transformation.apply(record), processingContext);
            if (result.failed()) {
                processingContext.report();
                return null;
            }
            record = result.result();
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -32,10 +37,26 @@ public TransformationChain(List<Transformation<R>> transformations) {
+        processingContext.position(Stage.TRANSFORMATION);
+        for (final Transformation<R> transformation : transformations) {
+            processingContext.executingClass(transformation.getClass());
+            final R current = record;
```
We don't need this, do we?

wicknicks replied:
lambdas need the inputs to be final variables. since we reassign the value or record, we can't make it final here. hence, the new (final) variable.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DLQReporter.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.Map;
+import java.util.concurrent.ThreadLocalRandom;
+
+public class DLQReporter implements ErrorReporter {
```
How about JavaDoc to explain what this does? I know it's internal, but it would help developers looking at this in the future.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DLQReporter.java:
```diff
@@ -0,0 +1,95 @@
+        ConsumerRecord<byte[], byte[]> originalMessage = context.sinkRecord();
+        int partition = ThreadLocalRandom.current().nextInt(numPartitions);
+        ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<>(config.topic(),
+                partition, originalMessage.key(), originalMessage.value());
```
I think it's a mistake to not include at least some information about why this record failed to be processed. I understand that we want to preserve the original formatting, but we could add headers with some relevant information. Otherwise, a user will have no clue why it was not processed.

wicknicks replied:
So we talked about this during the KIP discussions, and the general consensus was that we should take the simplest steps in this first version, and add such features in later versions, if there is demand for it.

When we specifically talked about context in headers, it was not clear what the header tags should be. Do we have some special system headers (which use `__` prefixes, for example `__error_stage` to denote a stage) which we can use here? We could always check for existing headers, but would there be any risk to overwriting existing headers? This also means that developers will have to clean up the records before reading them back into the sink task (which might not be needed if the corrections they had to make were to the connector pipeline as opposed to the records themselves).

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DLQReporter.java:
```diff
@@ -0,0 +1,95 @@
+                partition, originalMessage.key(), originalMessage.value());
+        this.kafkaProducer.send(producerRecord, (metadata, exception) -> {
+            if (exception != null) {
+                log.debug("Could not send object to Dead Letter Queue", exception);
```
Why is this debug?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,118 @@
+    public String toString() {
+        return "ProcessingContext{" +
+                "position=" + position +
+                ", klass=" + klass +
```
Why "klass" in the resulting string? Just use "class".

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/Result.java:
```diff
@@ -0,0 +1,50 @@
+public class Result<V> {
+
+    private final V result;
+    private final boolean success;
```
Why have a boolean that signals success, when `error != null` is equivalent to `success()`? Let's save the size when instantiating so many of these.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/Result.java:
```diff
@@ -0,0 +1,50 @@
+        this(null, false, error);
+    }
+
+    public Result(V result, boolean success, Throwable error) {
```
This constructor doesn't seem to be called, so why have it?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/Result.java:
```diff
@@ -0,0 +1,50 @@
+ */
+package org.apache.kafka.connect.runtime.errors;
+
+public class Result<V> {
```
How about making this class `final` to help the JIT inline/optimize the calls.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -32,10 +37,26 @@ public TransformationChain(List<Transformation<R>> transformations) {
+            if (result.success()) {
+                record = result.result();
+            } else {
+                processingContext.report();
```
Doesn't the executor call `context.report()` if there is a failure?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceExecutor.java:
```diff
@@ -0,0 +1,235 @@
+            markAsFailed();
+
+            Result<V> exResult = new Result<>(e);
+            context.result(exResult);
```
The previous two lines are unused if either of the next two if-blocks are entered. Could they be moved to after those if-blocks?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceExecutor.java:
```diff
@@ -0,0 +1,235 @@
+        public ToleranceType toleranceLimit() {
+            return "ALL".equals(getString(TOLERANCE_LIMIT).toUpperCase(Locale.ROOT)) ? ToleranceType.ALL
+                    : ToleranceType.NONE;
+        }
```
How about moving the parsing of the string into a `ToleranceType` into the enum itself as a static method?

rhauch reviewed (commented):
First pass, and this looks really good. A number of comments below.

wicknicks committed:
KAFKA-6738: Add some javadocs

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Address some more review comments

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Clean up code based on code reviews

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Return failure in execute() if ProcessingContext is in failed state

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Remove Result class and update some tests

Signed-off-by: Arjun Satish <arjun@confluent.io>

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:
```diff
@@ -231,6 +265,53 @@ public void execute() {
+
+        processingContext.setCurrentContext(Stage.HEADER_CONVERTER, headerConverter.getClass());
+        RecordHeaders headers = execute(() -> convertHeaderFor(record));
+        if (processingContext.failed()) {
```
I think these conditions are no longer required. Probably needs condition in the last step alone?

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -472,47 +500,64 @@ public String toString() {
+
+        processingContext.setCurrentContext(Stage.VALUE_CONVERTER, valueConverter.getClass());
+        SchemaAndValue valueAndSchema = execute(() -> valueConverter.toConnectData(msg.topic(), msg.value()));
+        if (processingContext.failed()) {
```
Assume these checks are no longer required and may be some condition to just create source record should be suffice?

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DLQReporter.java:
```diff
@@ -0,0 +1,118 @@
+
+        int partition = ThreadLocalRandom.current().nextInt(numPartitions);
+
+        ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<>(config.topic(),
```
We should include headers here. Also, not sure if we need to randomize the partition here.

wicknicks replied:
great catch! thanks!

mageshn reviewed (commented):
These are great changes. One  thought I had was if we could make ProcessingContext immutable in away that only the executor can change the workflow. This would mean that there will have to be public builder methods that help set or reset the context for each record. This will kind of encapsulate the context inside the executor.

wicknicks committed:
KAFKA-6738: Clean up unused variables in test

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6728: Add error handling tests for worker source task

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Remove unnecessary failure checks

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Produce headers into dead letter queue

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@rhauch, @mageshn, @kkonstantine this is ready for a review. thanks!

rhauch commented:
@wicknicks, @mageshn, @ewencp: The more I think about this, the more I think it is a problem to simply replicate the problem message _from potentially multiple topics_ into a new topic using the original message's partition and  _with no useful information_ about the original topic, partition, offset, connector name, task number, processing stage, and error.

First of all, the DLQ topic might have fewer partitions than the original topic, in which point this produce will fail. In fact, if the DLQ topic has a different number of partitions than the original topic, then the partitioning may be completely different.

Secondly, if the sink task is consuming from multiple topics, then we will be combining failed records from multiple topics.

Third, what does it mean to have a DLQ topic with multiple partitions? Do we really expect so many errors that we need to partition it at all? And having multiple partitions means we will not be able to maintain the total order of the errors.

Fourth, how do we actually _use_ the DLQ to do anything useful? We're missing all of the useful information about the connector, the connector task, the stage at which the exception happened, the timestamp at which the failure happened, the exception message, the original topic, the original partition, and the original offset. All we know is that there's this message (with key, value, and headers) that we couldn't process, but we know nothing else.

mageshn commented:
@rhauch I think we should include some of these traceability information like source topic, offset, partition, etc. in the message header. I'm all for just having 1 partition for the DLQ topic. I think @wicknicks had some concerns about durability there.

wicknicks commented:
@rhauch: We want to use the DLQ to simply carry the original message. Error context is logged as part of application logs with all the details you listed above. We want the DLQ to contain the original messages only. 

I agree that writing bad records from multiple topics into a single DLQ can get confusing to read from. But it is not clear how to handle this correctly. Adding metadata to the bad record corrupts the records (new headers [1]), writing to multiple DLQ topics (one for each source topic) might be wasteful. IMO, writing to a single DLQ is a simple option, which can start with (esp., given the feature freeze is in a couple of days), and improve with future releases.

BTW, the partition we produce to in the DLQ is not related to the original topic's partition. It's a randomly selected partition.

[1] adding new headers is a non-trivial change. we need to make sure they don't conflict with existing headers. and Connect might have to include something akin to "system headers", which might require a KIP of its own.

wicknicks commented:
I'm good with a single partition in the DLQ too (since we don't expect a demanding workload for errors). 

If we decide to write error metadata into Kafka, we should at least put it in a different topic. This way, we get original raw messages in the DLQ, and the error context and offset of the record in the DLQ in this error-context topic. People can choose to ignore the error context or use it depending on the situation.

Thoughts?

mageshn commented:
@wicknicks I think it would be a reasonable approach to say we will add headers with prefix DLQ_ . We might have to document this  and this would mean that if for some reason users use same prefix, they can potentially get overwritten. I really think the chances of this happening is really slim.

rayokota commented:
Perhaps for the DLQ we can also use 2 underscores as a prefix for metadata headers (topic, partition, offset, exception, etc), and maybe a config `errors.deadletterqueue.metadata.headers.enable`.  Just a thought.

rhauch commented:
@wicknicks wrote:

> We want to use the DLQ to simply carry the original message. Error context is logged as part of application logs with all the details you listed above. We want the DLQ to contain the original messages only.

Even if we were to include the error condition information as headers (we can simply use a more specific prefix as @mageshn mentioned), that DLQ topic may contain messages from multiple topics and multiple partitions, serialized in potentially multiple forms.

> I agree that writing bad records from multiple topics into a single DLQ can get confusing to read from. But it is not clear how to handle this correctly. Adding metadata to the bad record corrupts the records (new headers [1]), writing to multiple DLQ topics (one for each source topic) might be wasteful. IMO, writing to a single DLQ is a simple option, which can start with (esp., given the feature freeze is in a couple of days), and improve with future releases.

But users have to configure the DLQ on every connector where they want to use it. Seems like the only practical way to use this is to use a different topic for each connector. But they'll have to consume that DLQ topic using the same converter, and if the converter was the source of the problem then what?

In short, I'm struggling to see the usefulness of even having a DLQ. In particular, if the problem is an inability to deserialize/convert, then it's even harder to understand the value of replicating the original message. I guess when I envisioned a DLQ, I was thinking more of a "failure topic" that described all of the failures that my connector(s) have experienced, which optionally included the original information as `BYTES` fields within the larger failure message. Presumably, the failure message would include all of the context, including the converter that was used and the error, so that I could write an application that might at least let me understand the failures if not act on them and resolve them. Help me understand how the DLQ in the KIP might help me with that.

rhauch commented:
BTW, I'm fine with implementing the KIP as it has been approved, and I think this PR should do that. I just need to learn how useful it will be in practice.

wicknicks committed:
KAFKA-6738: Clean up DLQ setup

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Expose only OperationExecutor and its implementations

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@mageshn I updated this PR with your comment. ProcessingContext is now hidden behind OperationExecutor. All task classes only interact with the OperationExecutor or it's implementation now.

@rhauch: I do like @rayokota's approach. It gives us the best of both worlds. If users are ok with adding some headers, then they enable this flag, or if they want the records as it is, they don't override the default value. At the same time, the raw records remain as they were. I'll have to update the KIP and see if people are ok with those changes before implementing. Since feature freeze is approaching fast, I think it's better to defer the changes (should be relatively minor), to a smaller PR next week. Thanks!

wicknicks committed:
KAFAK-6738: msg does not need to be final

Signed-off-by: Arjun Satish <arjun@confluent.io>

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -27,15 +30,24 @@
+
+            // execute the operation
+            record = operationExecutor.execute(() -> transformation.apply(current), Stage.TRANSFORMATION, transformation.getClass());
+            if (operationExecutor.failed()) {
```
nit: probably not needed

wicknicks committed:
KAFKA-6738: Remove unecessary check for failure

Signed-off-by: Arjun Satish <arjun@confluent.io>

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetricsRegistry.java:
```diff
@@ -93,6 +94,14 @@
     public final MetricNameTemplate rebalanceTimeAvg;
     public final MetricNameTemplate rebalanceTimeSinceLast;
+    public final MetricNameTemplate recordProcessingFailures;
+    public final MetricNameTemplate recordProcessingErrors;
```
I think this is new and different from what's in the KIP? Do we need to update it there?

wicknicks replied:
The metrics here should be in sync with the KIP now.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetricsRegistry.java:
```diff
@@ -93,6 +94,14 @@
+    public final MetricNameTemplate recordProcessingErrors;
+    public final MetricNameTemplate recordsSkipped;
+    public final MetricNameTemplate retries;
+    public final MetricNameTemplate errorsLogged;
```
errors logged or failures logged? those are different based on the definitions for this KIP, right? again, does KIP need updating?

wicknicks replied:
actually, yes, these names looked more useful/natural. Would it be ok to update the KIP with the new names?

wicknicks replied:
since this https://www.mail-archive.com/dev@kafka.apache.org/msg88342.html hadn't got any response, I didn't bring up the metric name changes yet. Thoughts?

ewencp replied:
In general, we expect that there are some details that end up coming out only during PR review, so usually as long as it's not a major change, it's fine to update the KIP to match the actual implementation details. Still nice to send the note, but in the future I wouldn't get blocked on this (after all, we'll always have time during feature and code freeze to make fixes if we think we made a mistake).

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/OperationExecutor.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Execute a recoverable operation in the Connector pipeline.
+ */
+public interface OperationExecutor extends Configurable {
```
not a big deal here since it's all internal implementation details, but this seems like premature abstraction. there's only one implementation of this interface afaict, and we seem to frequently rely on methods that aren't in the interface here when we use the `RetryWithToleranceExecutor`.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -48,6 +57,10 @@ public void close() {
         }
     }
 
+    public void initialize(OperationExecutor operationExecutor) {
```
nit: for public apis, it's nicer to order the primary calls within the code file in the order they'd be invoked -- this should probably be moved up above `apply`.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -448,26 +453,72 @@ private WorkerTask buildWorkerTask(ConnectorConfig connConfig,
         if (task instanceof SourceTask) {
+            operationExecutor.setReporters(sourceTaskReporters(id, connConfig, errorHandlingMetrics));
             TransformationChain<SourceRecord> transformationChain = new TransformationChain<>(connConfig.<SourceRecord>transformations());
+            transformationChain.initialize(operationExecutor);
```
Why a separate `initialize` here instead of just making the `operationExecutor` a parameter to the constructor? Seems like it would be cleaner, make it possible to make it `final` in `TransformationChain`, etc.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -448,26 +453,72 @@ private WorkerTask buildWorkerTask(ConnectorConfig connConfig,
+        reporters.add(logReporter);
+
+        // check if topic for dead letter queue exists
+        String topic = connConfig.getString(DLQReporter.PREFIX + "." + DLQReporter.DLQ_TOPIC_NAME);
```
We should encode the full name of the config in `DLQReporter` instead of having this class construct it.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java:
```diff
@@ -283,6 +290,10 @@ protected void recordCommitFailure(long duration, Throwable error) {
         taskMetricsGroup.recordCommit(duration, false, error);
     }
 
+    protected <V> V executeOperation(Operation<V> operation, Stage stage, Class<?> klass) {
```
seems kind of weird to provide this helper wrapper given that `operationExecutor` is `protected` and therefore accessible in subclasses anyway, given the wrapper just passes parameters along directly. since we need to call the `sourceRecord` and `sinkRecord` methods anyway, can we just get rid of this?

kkonstantine replied:
I agree. Either the member field should be `private` or this method is not required.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DLQReporter.java:
```diff
@@ -0,0 +1,156 @@
+
+    private static final Logger log = LoggerFactory.getLogger(DLQReporter.class);
+
+    private static final int ADMIN_OPERATIONS_TIMEOUT_MILLIS = 10000;
```
Do we need this in addition to the regular AdminClient timeout configs, e.g. the request timeout that already exists (and is actually quite a bit larger than this)?

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DLQReporter.java:
```diff
@@ -0,0 +1,156 @@
+    private KafkaProducer<byte[], byte[]> kafkaProducer;
+    private ErrorHandlingMetrics errorHandlingMetrics;
+
+    static ConfigDef getConfigDef() {
```
why a static method rather than just defining the `ConfigDef` once statically? doesn't seem like we're doing anything like extending it when calling `getConfigDef`. also, `get` prefixes are non-kafkaesque (though not a bit deal here since it's internal)

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DLQReporter.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    public static DLQReporter createAndSetup(WorkerConfig workerConfig, ConnectorConfig connConfig, Map<String, Object> producerProps) {
+        String topic = connConfig.getString(PREFIX + "." + DLQ_TOPIC_NAME);
```
Doesn't look right -- `PREFIX` has a trailing dot, but this also adds another `.` before the DLQ topic name setting.

wicknicks replied:
great catch. thank you!

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DLQReporter.java:
```diff
@@ -0,0 +1,156 @@
+            if (!admin.listTopics().names().get(ADMIN_OPERATIONS_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS).contains(topic)) {
+                log.error("Topic {} doesn't exist. Will attempt to create topic.", topic);
+                int maxReplicationFactor = admin.describeCluster().nodes().get(ADMIN_OPERATIONS_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS).size();
+                int replicationFactor = Math.min(maxReplicationFactor, DLQ_MAX_DESIRED_REPLICATION_FACTOR);
```
We should be careful about an approach like this, this has been an issue elsewhere, e.g. using number of nodes available as the min replication factor for things like offsets topic turns out to cause serious problems -- if the topic happens to be created while only one node is up, it has an unacceptably low replication factor.

kkonstantine replied:
Also not too thrilled with this heuristic here. Not feeling too strong about this either.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ErrorHandlingMetrics.java:
```diff
@@ -0,0 +1,161 @@
+        Map<String, String> workerProps = new HashMap<>();
+        workerProps.put("key.converter", ByteArrayConverter.class.getName());
+        workerProps.put("value.converter", ByteArrayConverter.class.getName());
+        workerProps.put("internal.key.converter", ByteArrayConverter.class.getName());
```
why are the internal key/value converters also using `ByteArrayConverter`? especially given KIP-174, shouldn't these always be `JsonConverter` with settings to ensure correct schema-less behavior? (or post KIP-174, just omitted)

wicknicks replied:
@ewencp: the KIP had been updated to use producers to use not emit json converter. Also, this code is to create default ErrorHandlerMetrics, which is in turn used to construct the NOOP_EXECUTOR. These are default values to create a `StandaloneConfig`.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ErrorHandlingMetrics.java:
```diff
@@ -0,0 +1,161 @@
+        workerProps.put("value.converter", ByteArrayConverter.class.getName());
+        workerProps.put("internal.key.converter", ByteArrayConverter.class.getName());
+        workerProps.put("internal.value.converter", ByteArrayConverter.class.getName());
+        workerProps.put("offset.storage.file.filename", "/tmp/offsets.tmp");
```
How is this being used? Is it required? `/tmp` definitely won't always be accessible (either for read or write).

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/LogReporter.java:
```diff
@@ -0,0 +1,131 @@
+        this.id = id;
+    }
+
+    public LogReporter() {
```
This seems to only be used in test methods, maybe we could get rid of this if we provided `ConnectorTaskIds` in the tests instead of relying on parameterless constructors?

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/LogReporter.java:
```diff
@@ -0,0 +1,131 @@
+        this.id = new ConnectorTaskId("UNKNOWN", 0);
+    }
+
+    static ConfigDef getConfigDef() {
```
Same question here about just using a static ConfigDef instead of a static method.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceExecutor.java:
```diff
@@ -0,0 +1,314 @@
+                    errorHandlingMetrics.recordRetry();
+                } else {
+                    log.trace("Can't retry. start={}, attempt={}, deadline={}", startTime, attempt, deadline);
+                    context.error(e);
```
there's a `recordRetry()` above, no `recordError()` needed here, or is it guaranteed by the `context.error()`? if so, asymmetry is not clear on first reading...

wicknicks replied:
`recordError()` is called when all retries are attempted (yes, it uses the value of `context.error()`): https://github.com/wicknicks/kafka/blob/d2b93c5117fc960023a4866266187988d61baf9b/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceExecutor.java#L141.

wicknicks committed:
KAFKA-6738: Remove OperationExecutor interface

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Remove empty constructor for LogReporter

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Set OperationExecutor in TransformationChain constructor

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@ewencp: let me know what you think about the KIP changes (metrics and [1]). Depending on that, I'll make the changes in the code. Thanks very much!

[1] https://www.mail-archive.com/dev@kafka.apache.org/msg88342.html

wicknicks committed:
KAFKA-6738: Update metric names

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Remove dependency of ConnectMetrics on WorkerConfig

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@ewencp, @rhauch, @mageshn I think I've addressed all your comments. Could you please take a look at this PR? Thanks!

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:
```diff
@@ -231,6 +255,43 @@ public void execute() {
+        try {
+            return task.poll();
+        } catch (RetriableException e) {
+            log.error("Encountered retriable error on polling task", e);
```
Given that this log message is an error I think it should be more informative about what is going to happen. Also, should this be a warning instead, given that it's not fatal. Just wondering

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DLQReporter.java:
```diff
@@ -0,0 +1,156 @@
+ * queue topic. By default, the topic name is not set, and if the connector config doesn't specify one, this
+ * feature is disabled.
+ */
+public class DLQReporter implements ErrorReporter {
```
I know that elsewhere, more than 3 consecutive caps is an issue for checkstyle. Is`DlqReporter` that much worse and worth risking a potential exception in checkstyle rules (even in the future)? Also seems to be the only class with DLQ in the name.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DLQReporter.java:
```diff
@@ -0,0 +1,156 @@
+     *
+     * @param kafkaProducer a Kafka Producer to produce the original consumed records.
+     */
+    DLQReporter(KafkaProducer<byte[], byte[]> kafkaProducer) {
```
Relatively easy to miss that this is package private. Especially with a javadoc preceding the constructor. Probably better to make the intended use more explicit by using `protected`. Also if this is not `private` for testing we should add the usual `//visible for testing` comment. (again javadoc is slightly misleading here and is probably more appropriate in the factory method above - which also is public instead).

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DLQReporter.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    /**
+     * @param context write the record from {@link ProcessingContext#consumerRecord()} into the dead letter queue.
```
Not a big fan of javadoc sections that include text only for parameters/return values. The redundancy is not that big if one writes `@param 'the thing' (implying 'the thing described above'). Especially here, without a comma between context and write (which is correct since the separator of the param name and the description is white space), the description doesn't read very well.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ErrorHandlingMetrics.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    /**
+     * increment the number of failed operations (retriable and non-retriable).
```
Are we against `caps` on javadocs :) ? I wouldn't mind, if this is not in striking difference from other javadocs.

wicknicks replied:
fixed :)

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/LogReporter.java:
```diff
@@ -0,0 +1,127 @@
+                .define(LOG_INCLUDE_MESSAGES, ConfigDef.Type.BOOLEAN, LOG_INCLUDE_MESSAGES_DEFAULT, ConfigDef.Importance.MEDIUM, LOG_INCLUDE_MESSAGES_DOC);
+    }
+
+    public void report(ProcessingContext context) {
```
Missing `@Override` here.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/Operation.java:
```diff
@@ -0,0 +1,33 @@
+public interface Operation<V> {
+
+    /**
+     * evaluate the operation
```
Pls use caps :) (see your `ErrorReporter` above).

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,189 @@
+    private SourceRecord sourceRecord;
+
+    /**
+     * the following fields need to be reset every time a new record is seen.
```
For a nice example where caps make sense see right below, where two sentences are included.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,189 @@
+    private Throwable error;
+
+    /**
+     * Only one instance of this class is meant to exist per task in a JVM. This method resets the internal fields
```
This comment sounds like it should definitely be included in the top level javadoc of the class.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    /**
+     * set the stage in the connector pipeline which is currently executing.
```
Reading up to here, I'm convinced caps or no caps is random :)

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,189 @@
+        return this.error() != null;
+    }
+
+    public void setReporters(Collection<ErrorReporter> reporters) {
```
It's not an override and misses javadoc

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,189 @@
+     * @return the number of attempts made to execute the current operation.
+     */
+    public int attempt() {
+        return this.attempt;
```
nit: as in `position` above, `this` is not required

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,189 @@
+     * @return the source record being processed.
+     */
+    public SourceRecord sourceRecord() {
+        return this.sourceRecord;
```
nit: as in `position` below, `this` is not required

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,189 @@
+     * @return the record consumed from Kafka. could be null
+     */
+    public ConsumerRecord<byte[], byte[]> consumerRecord() {
+        return this.consumedMessage;
```
nit: as in `position` below, `this` is not required

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,189 @@
+     * @return the class which is going to execute the current operation.
+     */
+    public Class<?> executingClass() {
+        return this.klass;
```
nit: as in `position` above, `this` is not required

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceExecutor.java:
```diff
@@ -0,0 +1,312 @@
+ * then it is wrapped into a ConnectException and rethrown to the caller.
+ * </p>
+ */
+public class RetryWithToleranceExecutor {
```
Glad to see `OperationExecutor` interface was removed. Haven't read all the comments and I'm not sure what was the reason. But I find the `Executor` suffix misleading if it doesn't implement the `Executor` interface (which is a bit specific in the java library). Should we consider a removal from here too? Not too strong about this and the problem is that the executor name is everywhere in the code (but not in the keep). Still a bit misleading because this class is not an executor in the narrow sense.

wicknicks replied:
Any thoughts on what else this class could be called? I understand the conflict with java's Executor? But other words (for example, evaluator, handler) didn't really fit the bill too. In a sense this class executes `Operations`, whereas Java's executors execute `Runnable`. In that sense, there is some similarities. Since this name was chosen when we had an interface, should we simply call this class `OperationExecutor`?

kkonstantine replied:
`RetryWithToleranceOperator`. Basically, an operation executor is ... an operator. The word has somehow specific meaning too, but in a different context, so maybe it can't be confused that easily. 

Using Operator instead of Executor could leave most of your javadocs unchanged and could also have an `execute` method (not too bad). Just a suggestion.

wicknicks replied:
Renamed the class. I suppose the biggest challenge to calling it an operator was that usually multiple operators typically exist/work together (arithmetic or query operators, for example). It's a bit weird to have a single operator. But, as you said, it does reduce confusion.

kkonstantine reviewed (commented):
Thanks @wicknicks. Great addition! 
Late in the party, I have some minor comments. The most global is related to some javadoc consistency (seems little effort too). Also glad that `OperationExecutor` was removed. `Executor` is a bit misleading if it doesn't implement the jvm interface. 

Running out of battery, hope my comments help a bit. Overall great work.

wicknicks committed:
KAFKA-6738: Address reviewer comments

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Set the replication factor for the dead letter queue topic

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Address reviewer comments and fix bugs

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Captilize some javadoc

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Clean up javadocs and code style

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@rhauch, @mageshn, @ewencp, @kkonstantine I think I've addressed all your comments. Please take a look at this. If possible, let's try to fix any pending issues before the feature freeze. Thanks very much!

wicknicks committed:
KAFKA-6738: Remove trailing character from PREFIX

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6738: Rename RetryWithToleranceExecutor with RetryWithToleranceOperator

Signed-off-by: Arjun Satish <arjun@confluent.io>

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetricsRegistry.java:
```diff
@@ -294,6 +303,28 @@ public ConnectMetricsRegistry(Set<String> tags) {
+        dlqProduceRequests = createTemplate("deadletterqueue-produce-requests", TASK_ERROR_HANDLING_GROUP_NAME,
+                "The number of times a report request was made to the dead letter queue reporter.", taskErrorHandlingTags);
+        dlqProduceFailures = createTemplate("deadletterqueue-produce-failures", TASK_ERROR_HANDLING_GROUP_NAME,
+                "The number of times a report request to the dead letter queue reporter failed.", taskErrorHandlingTags);
```
How about `The number of failed writes to the dead letter queue."?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetricsRegistry.java:
```diff
@@ -294,6 +303,28 @@ public ConnectMetricsRegistry(Set<String> tags) {
+        errorsLogged = createTemplate("total-errors-logged", TASK_ERROR_HANDLING_GROUP_NAME,
+                "The number of times, error context was logged due to an error.", taskErrorHandlingTags);
+        dlqProduceRequests = createTemplate("deadletterqueue-produce-requests", TASK_ERROR_HANDLING_GROUP_NAME,
+                "The number of times a report request was made to the dead letter queue reporter.", taskErrorHandlingTags);
```
How about "The number of attempted writes to the dead letter queue."?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetricsRegistry.java:
```diff
@@ -294,6 +303,28 @@ public ConnectMetricsRegistry(Set<String> tags) {
+        retries = createTemplate("total-retries", TASK_ERROR_HANDLING_GROUP_NAME,
+                "The number of operations retried.", taskErrorHandlingTags);
+        errorsLogged = createTemplate("total-errors-logged", TASK_ERROR_HANDLING_GROUP_NAME,
+                "The number of times, error context was logged due to an error.", taskErrorHandlingTags);
```
How about "The number of errors that were logged."?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetricsRegistry.java:
```diff
@@ -294,6 +303,28 @@ public ConnectMetricsRegistry(Set<String> tags) {
+        dlqProduceFailures = createTemplate("deadletterqueue-produce-failures", TASK_ERROR_HANDLING_GROUP_NAME,
+                "The number of times a report request to the dead letter queue reporter failed.", taskErrorHandlingTags);
+        lastErrorTimestamp = createTemplate("last-error-timestamp", TASK_ERROR_HANDLING_GROUP_NAME,
+                "The epoch timestamp when this task encountered an error.", taskErrorHandlingTags);
```
How about "The epoch timestamp when this task last encountered an error."?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -26,16 +28,26 @@
+
+    public TransformationChain(List<Transformation<R>> transformations, RetryWithToleranceOperator retryWithToleranceOperator) {
         this.transformations = transformations;
+        this.retryWithToleranceOperator = retryWithToleranceOperator;
```
Can the second parameter be null?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java:
```diff
@@ -0,0 +1,331 @@
+     * @param <V> return type of the result of the operation.
+     * @return result of the operation
+     */
+    public <V> V execute(Operation<V> operation, Stage stage, Class<?> executingClass) {
```
Have you thought about rearranging the parameters so that the operation is last? It's the only parameter that might be a lambda, and having it last would be more readable.

ewencp replied:
In this case everything is quite readable since all the things we're delegating to are super short method calls, I found the code that invokes this quite readable (but of course that's subjective)

wicknicks replied:
yeah, I would prefer to keep the operation as the first argument, since it's the most important one. Generally, prefer to order arguments by importance. Also, I found it more readable too.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -26,16 +28,26 @@
+            final R current = record;
+
+            // execute the operation
+            record = retryWithToleranceOperator.execute(() -> transformation.apply(current), Stage.TRANSFORMATION, transformation.getClass());
```
This would be more readable if the operator parameter is last, since that's the one parameter that's likely to be a lambda.

wicknicks replied:
responded below.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -448,26 +453,71 @@ private WorkerTask buildWorkerTask(ConnectorConfig connConfig,
+
+    private List<ErrorReporter> sourceTaskReporters(ConnectorTaskId id, ConnectorConfig connConfig,
+                                                      ErrorHandlingMetrics errorHandlingMetrics) {
+        ArrayList<ErrorReporter> reporters = new ArrayList<>();
```
Minor: use the most generic type possible:

    List<ErrorReporter> reporters = new ArrayList<>();

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -477,42 +496,49 @@ private void convertMessages(ConsumerRecords<byte[], byte[]> msgs) {
-                    this, msg.topic(), msg.partition(), msg.offset(), timestamp, keyAndSchema.value(), valueAndSchema.value());
-            SinkRecord transRecord = transformationChain.apply(origRecord);
+
+            retryWithToleranceOperator.consumerRecord(msg);
```
Minor: might this be more readable if this field is just `operator`?

ewencp replied:
I actually like it as is because `operator` is super generic and I have no idea what operation it is doing. Since this operator is a wrapper that does retries and applies tolerance, I think the full name is actually clearer.

wicknicks replied:
yeah, I agree with @ewencp. I found `operator` too abstract, and found it easier to read the statement with the full name.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -477,42 +496,49 @@ private void convertMessages(ConsumerRecords<byte[], byte[]> msgs) {
             } else {
-                log.trace("{} Transformations returned null, so dropping record in topic '{}' partition {} at offset {} and timestamp {} with key {} and value {}",
-                        this, msg.topic(), msg.partition(), msg.offset(), timestamp, keyAndSchema.value(), valueAndSchema.value());
+                log.trace("{} Converters and transformations returned null, so dropping record in topic '{}' partition {} at offset {}",
```
Maybe "... returned null, possibly because of too many retries, so ..."?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -477,42 +496,49 @@ private void convertMessages(ConsumerRecords<byte[], byte[]> msgs) {
+        Headers headers = retryWithToleranceOperator.execute(() -> convertHeadersFor(msg), Stage.HEADER_CONVERTER, headerConverter.getClass());
+
+        if (retryWithToleranceOperator.failed()) {
+            return null;
```
Does this check need to be performed after each call to `operator.execute(...)`, rather than just once after the 3 calls?

ewencp replied:
@wicknicks I think this is technically true. `execute` actually does log enough info to figure out what's going on, as long as you're at trace level. But since the rest of the logging happens outside that context, you can end up in a somewhat unfortunate situation.

For the `LogReporter` this isn't terrible -- you log more errors and have to know to go look for the earliest message, but it just logs more info. For the DLQ, however, the behavior if we don't check on every operation is worse -- we'll duplicate errors to the DLQ as each `execute` call does its `report`, which would write a new message to the DLQ for the DLQ reporter. More confusing, it will report anywhere between 1-3 copies since it depends on what stage failed.

mageshn replied:
Operator code ensure that a stage is not executed if the pervious stage had failed. This is the first check in the execute().  The confusion arises because of the check for failed() in the worker tasks. Ideally, it would be cleaner to check the objects to be not null than checking failed().

wicknicks replied:
Apologies for the confusion. If a prior operation has failed, then any call to `execute` is a noop. It returns immediately (https://github.com/apache/kafka/pull/5065/files#diff-7981e38304219f029769a16784286d77R131). The javadoc for `execute` does mention this.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:
```diff
@@ -231,6 +255,44 @@ public void execute() {
+                Stage.VALUE_CONVERTER, valueConverter.getClass());
+
+        if (retryWithToleranceOperator.failed()) {
+            return null;
```
Again, should this check be made after each call to `operator.execute(...)`? I guess I'm just wondering if we might go beyond the max # of allowed failures with this. Perhaps that's not really a big deal, though.

mageshn replied:
Operator code ensure that a stage is not executed if the pervious stage had failed. This is the first check in the execute(). The confusion arises because of the check for failed() in the worker tasks. Ideally, it would be cleaner to check the objects to be not null than checking failed().

wicknicks replied:
replied to this above. thanks, @mageshn!

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java:
```diff
@@ -60,11 +61,14 @@
     private volatile boolean stopping;   // indicates whether the Worker has asked the task to stop
     private volatile boolean cancelled;  // indicates whether the Worker has cancelled the task (e.g. because of slow shutdown)
 
+    protected final RetryWithToleranceOperator retryWithToleranceOperator;
```
I mentioned this elsewhere, but now "operator" seems to be fairly clear, so perhaps we could make the code more readable by just calling this field `operator`. WDYT?

wicknicks replied:
responded below.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DeadLetterQueueReporter.java:
```diff
@@ -0,0 +1,153 @@
+ * Write the original consumed record into a dead letter queue. The dead letter queue is a Kafka topic located
+ * on the same cluster used by the worker to maintain internal topics. Each connector gets their own dead letter
+ * queue topic. By default, the topic name is not set, and if the connector config doesn't specify one, this
+ * feature is disabled.
```
What happens if multiple connectors are configured to use the same DLQ topic? Is that a problem, or just not recommended? Do we enforce it?

ewencp replied:
we shouldn't enforce. un-schema'd formats like json might be fine sharing the topic. we should continue to discuss what additional metadata we add to the DLQ messages, but with the right metadata, there's really no problem sharing the same queue if someone wants that simplicity. even in decoding avro, they can still decode based on schema ID, they just need enough context to figure out what to do with it

rhauch replied:
Right, so we should just reword “Each connector gets their own dead letter queue topic” to be more clear, so something like “Each connector’s dead letter queue is usually written to a different topic”. Not super important, but I’m just looking to eliminate potential usability issues.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DeadLetterQueueReporter.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    @Override
+    public void setMetrics(ErrorHandlingMetrics errorHandlingMetrics) {
```
AK convention is to not use `set` setters or `get` getters.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DeadLetterQueueReporter.java:
```diff
@@ -0,0 +1,153 @@
+        }
+
+        ProducerRecord<byte[], byte[]> producerRecord;
+        if (originalMessage.timestamp() > 0) {
```
This seems error prone to be checking for this, rather than using the ConsumerRecord's timestamp type.

ewencp replied:
Agreed here, can we use https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/record/RecordBatch.java#L47 here to be specific to the one sentinel value we have? That would also make it robust to changes for https://cwiki.apache.org/confluence/display/KAFKA/KIP-228+Negative+record+timestamp+support (maybe not robust to every possible fix, but at least for ones that just reserve that one timestamp value as a sentinel value).

rhauch replied:
Yeah, that works, too, and is more align with the current code.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ErrorReporter.java:
```diff
@@ -0,0 +1,38 @@
+    /**
+     * @param errorHandlingMetrics metrics for error handling
+     */
+    void setMetrics(ErrorHandlingMetrics errorHandlingMetrics);
```
This style of setters is not AK convention.

ewencp replied:
nit: in kafka we normally don't use `get`/`set` prefixes

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ErrorReporter.java:
```diff
@@ -0,0 +1,38 @@
+    /**
+     * Report an error.
+     *
+     * @param context the processing context.
```
Nit: can the `context` be null, or can an implementation assume it will never be null?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ErrorReporter.java:
```diff
@@ -0,0 +1,38 @@
+    void report(ProcessingContext context);
+
+    /**
+     * @param errorHandlingMetrics metrics for error handling
```
Nit: will this ever be null? Would be nice to know in an the implementation if it can assume it's never null.

Also, need a description of the method.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/LogReporter.java:
```diff
@@ -0,0 +1,131 @@
+import java.util.Map;
+
+/**
+ * Log error context with application logs.
```
Nit:  reword to avoid "log" being ambiguous as verb or noun: "Writes errors and their context to application logs."

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/LogReporter.java:
```diff
@@ -0,0 +1,131 @@
+    public static final String PREFIX = "errors.log";
+
+    public static final String LOG_ENABLE = "enable";
+    public static final String LOG_ENABLE_DOC = "Log the error context along with the other application logs.";
```
Might be more useful if this explained what an "error context" is. Something like:

    Log to application logs the errors and the information describing where they occurred.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/LogReporter.java:
```diff
@@ -0,0 +1,131 @@
+    public static final boolean LOG_ENABLE_DEFAULT = false;
+
+    public static final String LOG_INCLUDE_MESSAGES = "include.messages";
+    public static final String LOG_INCLUDE_MESSAGES_DOC = "Include the Connect Record which failed to process in the log.";
```
Maybe:

  Include in the log the Connect key, value, and other details of records that resulted in errors and failures.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/LogReporter.java:
```diff
@@ -0,0 +1,131 @@
+            builder.append(", msg.partition=").append(msg.partition());
+            builder.append(", msg.offset=").append(msg.offset());
+            builder.append(", msg.timestamp=").append(msg.timestamp());
+            builder.append(", msg.timestampType=").append(msg.timestampType());
```
Do we need to prefix these with `msg.`?

Also, it might be nice to control the message format when message keys and values are to be included. I know that's more code, but it could be made to be similar to the other option.

rhauch replied:
Is there any way to put this into methods on the `ProcessingContext`, like `toString(boolean includeRecord)`? That way, its `toString()` method could really just be `toString(false)`. Not only is it a bit more testable, the logic of this reporter is more readable and focused on writing to the log.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/Operation.java:
```diff
@@ -0,0 +1,34 @@
+     * @return the result of the operation.
+     * @throws Exception if underlying operation fails.
+     */
+    V apply() throws Exception;
```
Why not just use `java.util.concurrent.Callable` here?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,199 @@
+
+/**
+ * Contains all the metadata related to the currently evaluating operation. Only one instance of this class is meant
+ * to exist per task in a JVM. The {@link #reset} should be called before starting operations on a new record.
```
Remove the last sentence, since the `reset` method is private, and looks like it's handle automatically.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,199 @@
+     *
+     * @param reporters the error reporters (should not be null).
+     */
+    public void setReporters(Collection<ErrorReporter> reporters) {
```
Doesn't follow AK setter convention. Any reason to maybe set this in an `initialize` method? Seems like it should be called up front, rather than repeatedly. Anything else that should be set only once? Removing unnecessary setters might help with the usability.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java:
```diff
@@ -0,0 +1,331 @@
+/**
+ * <p>
+ * Attempt to recover a failed operation with retries and tolerance limits.
+ * </p>
```
Nit: no paragraph tags around the first paragraph. And never use `</p>`, since JavaDoc doesn't use them (they're an HTML thing).

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java:
```diff
@@ -0,0 +1,331 @@
+    private static final Logger log = LoggerFactory.getLogger(RetryWithToleranceOperator.class);
+
+    public static final String RETRY_TIMEOUT = "retry.timeout";
+    public static final String RETRY_TIMEOUT_DOC = "The maximum number of retries before failing an operation";
```
This doc seems wrong for `retry.timeout`. Or is this really a retry limit? Either way, might want to mention the constraints (e.g., "Must be a positive number.") that are enforced by the config.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java:
```diff
@@ -0,0 +1,331 @@
+     *
+     * @param reporters the error reporters (should not be null).
+     */
+    public void setReporters(List<ErrorReporter> reporters) {
```
Setter methods again.

rhauch reviewed (commented):
Lots of minor suggestions, and a few moderate ones. Looking good, though!

rhauch commented:
@wicknicks, all of my comments are superficial and easy to fix. Once they are, I'm happy with this.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -26,16 +28,26 @@
+    private final RetryWithToleranceOperator retryWithToleranceOperator;
 
     public TransformationChain(List<Transformation<R>> transformations) {
+        this(transformations, RetryWithToleranceOperator.NOOP_OPERATOR);
```
nit: we seem to be adding a bunch of constructors for classes like these that seem to really only be there for tests, and don't simplify things much -- basically just like 3 or 4 lines of slightly repetitive stuff. I prefer to avoid doing this since it starts to make it unclear what the *correct* constructor is; sometimes it is unavoidable, but then at least we should add a `for testing only` or similar comment. Even better is if we can keep utilities that are intended just for testing to the test jar. Here that's a bit more annoying, so leaving `NOOP_OPERATOR` in the main jar is fine, but not ideal.

The reason this stood out to me is because it looks like we're also creating metrics for a bogus connector, but since it happens via static initialization, it's not avoidable.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java:
```diff
@@ -0,0 +1,199 @@
+     * @param stage the stage
+     * @param klass the class which will execute the operation in this stage.
+     */
+    public void setCurrentContext(Stage stage, Class<?> klass) {
```
nit: in kafka we usually don't use `get`/`set` prefixes

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java:
```diff
@@ -0,0 +1,331 @@
+            delay = deadline - time.milliseconds();
+        }
+        log.debug("Sleeping for {} millis", delay);
+        time.sleep(delay);
```
Hmm, this doesn't need to block merging this, but we should think carefully about doing delay this way. The rest of Connect avoids trying to rely on Java's `interrupt` behavior because it's not really a reliable way to *actually* interrupt threads, and in a system where there are pluggable components that are allowed to block indefinitely, relying on functionality that most Java developers don't understand well probably isn't going to work all that well.

It may not have actually gotten to a KIP, but there was at least some discussion on a JIRA somewhere about making connect perform interrupts in addition to the basic task `stop()` calls it already does, but it doesn't currently do this. For anything that can end up with pretty long sleep periods, we should try to make sure there's a good way of interrupting it and moving on (e.g. so rebalances wouldn't get delayed because there's a connector that's encountering errors). At a minimum, since we don't do interrupts currently, I think we wouldn't interrupt this code currently. The other approach we use elsewhere is to `wait` on a monitor so we can set a flag and interrupt with `notify` and have it bail out immediately.

wicknicks replied:
Interesting. Yeah, happy to have any discussions on what is a better way of doing this!

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperatorTest.java:
```diff
@@ -0,0 +1,308 @@
+        retryWithToleranceOperator.markAsFailed();
+        assertFalse("should not tolerate any errors", retryWithToleranceOperator.withinToleranceLimits());
+
+        retryWithToleranceOperator = new RetryWithToleranceOperator();
```
nit: not a big deal here, but for unit tests I think given the very low overhead it is better to separate out each of the cases into their own test as it can help make it more quickly obvious if issues are with a specific case or if it affects multiple cases.

ewencp committed:
Address a bunch of review comments and reduce extraneous public constructors only needed for tests.

kkonstantine reviewed (approved):
My comments have been addressed and I'm covered by the rest of the comments from everyone else. LGTM by me! Nice once @wicknicks

wicknicks committed:
KAFKA-6738: Address review comments

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DeadLetterQueueReporter.java:
```diff
@@ -0,0 +1,154 @@
+        Long timestamp = null;
+        if (originalMessage.timestampType() == TimestampType.LOG_APPEND_TIME || originalMessage.timestampType() == TimestampType.CREATE_TIME) {
+            timestamp = originalMessage.timestamp();
+        }
```
@rhauch, @ewencp: I hope this is a better way of setting the timestamp (if I understood your comment correctly). Please let me know if something else was required. thanks!

wicknicks commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/Operation.java:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @param <V> return type of the result of the operation.
+ */
+public interface Operation<V> extends Callable<V> {
```
@rhauch decided to make `Operation` extend `Callable`. Since we have been referring to an operation as a unit of work in the connect pipeline, I think there is value in having a specialized entity called `Operation` in the code. Also, your point is valid, hence making it extend `Callable`, and renaming the `apply` method to `call`. Thanks!

wicknicks commented:
@rhauch I think I've addressed all your comments. Many thanks! Really appreciate you taking the time to take multiple passes at this.

rhauch reviewed (approved):
LGTM. Nice work, @wicknicks!

ewencp reviewed (approved):
LGTM, merging to trunk for 2.0.0
